netket/logging/hdf5_log.py

```diff
@@ -0,0 +1,184 @@
+    - arrays are stored in the same way, but with values having shape (n_steps, *array_shape),
+    - netket.stats.Stats are stored as a group containing each field (Mean, Variance, etc...) as a separate dataset.
+
+    Importantly, each group has a dataset `iters`, which tracks the iteration number of the logged quantity.
+
+    If the model state is serialized, then it is serialized as a dataset in the group `parameters/`.
+    The target of the serialization are the parameters of the variational state,
+    which are stored in a dataset `parameters/value` as a flattened array.
+
+    Data can be deserialized by calling :code:`f = h5py.File(filename, 'r')` and
+    inspecting the datasets as a dictionary, i.e. :code:`f['data/energy/Mean']`
+    
+    .. note::
+        The API of this logger is covered by our Semantic Versioning API guarantees. However, the structure of the
+        logged files is not, and might change in the future while we iterate on this logger. If you think that we
+        could improve the output format of this logger, please open an issue on the NetKet repository and let us
+        know.
+    
+    """
+
+    def __init__(
+        self,
+        output_prefix: str,
+        mode: str = "write",
+        save_params: bool = True,
+        save_params_every: int = 1,
+    ):
+        """
+        Construct a HDF5 Logger.
+
+        Args:
+            output_prefix: the name of the output files before the extension
+            mode: Specify the behaviour in case the file already exists at this
+                output_prefix. Options are
+                - `[w]rite`: (default) overwrites file if it already exists;
+                - `[x]` or `fail`: fails if file already exists;
+            save_params: bool flag indicating whether variables of the variational state
+                should be serialized at some interval
+            save_params_every: every how many iterations should machine parameters be
+                flushed to file
+        """
+        import h5py
+
+        super().__init__()
+
+        if not ((mode == "write") or (mode == "append") or (mode == "fail")):
+            raise ValueError(
+                "Mode not recognized: should be one of `[w]rite`, `[a]ppend` or"
+                "`[x]`(fail)."
+            )
+        mode = _mode_shorthands[mode]
+
+        if output_prefix.endswith(".hdf5"):
+            output_prefix = output_prefix[:-5]
+
+        file_exists = os.path.exists(output_prefix + ".hdf5")
+
+        if file_exists and mode == "x":
+            raise ValueError(
+                "Output file already exists. Either delete it manually or"
+                "change `output_prefix`."
+            )
+
+        dir_name = os.path.dirname(output_prefix)
+        if dir_name != "":
+            os.makedirs(dir_name, exist_ok=True)
+
+        self._prefix = output_prefix
+        self._file_mode = mode
+        self._file_name = self._prefix + ".hdf5"
```

Comment:
I think `.h5` is the more common suffix for HDF5 files than `.hdf5`. Since both are fine though, maybe we should just let users decide what to use and not add an extension? (The logger only creates one file anyways, it is okay if the user defined prefix already contains the file extension.)

In that case, I'd just remove the `self._prefix` field and only store `self._file_name`

Reply:
It might be useful to have a standard file extension for checkpointing at some point later. Maybe we can jsut stick with the most common one and ditch `.hdf5` in favor of `.h5`?